zip-0320.rst

```diff
@@ -5,7 +5,347 @@
+Traceable Unified Addresses
+---------------------------
+
+A Traceable Unified Address is a reencoding of a transparent Zcash address into
+a Unified Address [#zip-0316-unified-addresses]_. 
+
+Motivations for Alternative 2
+-----------------------------
+
+Traceable Unified Addresses fit into the existing Zcash Unified Address
+ecosystem. Existing Consumers that support Unified Addresses will automatically
+be able to recognize a Traceable Unified Address as a valid Zcash address, but
+will not be able to send to that address unless they update their code to
+understand the new Receiver Typecode defined in this ZIP. Even in the case that
+Traceable Receivers are not understood by the sending wallet, a Unified
+Address-supporting wallet will be able to automatically provide good error
+messages for their users to indicate that the wallet needs to be updated to
+understand and send to these addresses.
+
+In addition, by integrating with the Unified Address framework, it becomes
+possible for the addresses being generated to include extra metadata, in
+particular, metadata items such as an Address Expiry Height or Address Expiry
+Date [#zip-0316-address-expiry]_ may be included. For exchange use cases such
+as Binance's, it is useful to ensure that an address provided to a user has a
+limited utility life, such that after expiration the user must obtain a new
+address in order to be able to continue to send funds
+[#binance-address-expiry]_.
+
+Specification (Alternative 2)
+-----------------------------
+
+Upon activation of this ZIP, the section `Encoding of Unified Addresses` of ZIP
+316 [#zip-0316-unified-addresses]_ will be modified to define a new
+Traceable Receiver Type having typecode :math:`\mathtt{0x04}`, the value of
+which MUST be the 20-byte **validating key hash** of a Zcash P2PKH Address as
+defined in [#protocol-transparentaddrencoding]_.
+
+The "Requirements for both Unified Addresses and Unified Viewing Keys" section
+of ZIP 316 [#zip-0316-unified-requirements]_ will be modified as follows —
+the text::
+
+  A Unified Address or Unified Viewing Key MUST contain at least one
+  shielded Item (Typecodes :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`).
+  The rationale is that the existing P2SH and P2PKH transparent-only
+  address formats, and the existing P2PKH extended public key format,
+  suffice for representing transparent Items and are already supported
+  by the existing ecosystem.
+
+will be replaced by::
+
+  A Unified Address MUST contain at least one Receiver and any number
+  of Metadata Items. The selection of Receivers is further restricted 
+  such that a Unified Address MUST **either** contain at least one shielded
+  Receiver (Typecodes :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`), OR 
+  MUST contain **only** a Traceable Address Receiver (Typecode
+  :math:`\mathtt{0x04}`).
+
+  A Unified Viewing Key MUST contain at least one shielded Item (Typecodes
+  :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`). 
+
```

Comment:
I think it's also necessary to add "Traceable P2PKH Receiver" to the ZIP 316 priority list. You might think that (after applying the suggestion above) it's not necessary to define its priority, because for any UA that has a Traceable P2PKH Receiver, it will be the only Receiver. However, ZIP 316 defines the "preferred" ordering like this:

> We say that a Receiver Type is “preferred” over another when it appears earlier in this Priority List (as potentially modified by experiments).

which means that it is not well-defined to specify that the most-preferred Receiver is used unless all Received types have a priority. Also, implementations will in practice give all types a priority because that's the simplest way to implement it.

Reply:
Hm. So does that mean that a Traceable P2PKH Receiver has the highest priority? I think that I would prefer to mention Traceable P2PKH Receivers in the section where priorities are defined, but instead of assigning a priority clarify that the priority of a Traceable P2PKH Receiver is undefined as it must never be present in an address with any other receivers.

Sum type, rather than picking an "unused" element of the domain, you know?

Reply:
It doesn't matter as long as the Receiver selection is well-defined.